044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,173 @@
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
+This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
+means the nodes will be able to communicate even if they are currently marked as not ready.
+
+### Impact on the KafkaRoller
+
+The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled. 
+This relies on the fact that currently the pod being ready implies that the BrokerState is >= RUNNING.
+Additionally, it does not take into account the state of other broker pods.
+The move to KRaft mode introduces some new requirements in terms of when a pod should or should not be rolled.
+This combined with the varying readiness checks means we need to change the way the KafkaRoller works:
+
+* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
+* If more than one controller pod has not become ready, KafkaRoller should try to role the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to STARTED and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to STARTED, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
```

Comment:
AFAIU, the above four lines sound to be an explanation of why the Readiness for "Broker only mode" and "Combined mode" is `the BrokerState metric >=3 && != 127`. If I got it right, would it make sense to provide this explanation in the previous section?

Reply:
The above four lines explains why the KafkaRoller should verify that the controller pods are not stuck in pending before waiting for broker pods to be ready. So I think it is right that it sits here. This proposal states that the KafkaRoller should stop relying on readiness to determine whether to roll pods and use it's own checks. So I think having this further up where we are discussing the readiness checks would muddle those back up again.